### ISSUE TITLE
Animals do not stay in pen

### ISSUE THREAD

**Issue by jungerm2** [cuberite/cuberite#3215]
Client version:   1.9.4
Commit id:        30e6425ed3e2d2c458158c9b8dd0838b60637c32
### Expected behavior

When making a cattle farm, the cows, and other animals are expected to stay in the pen (made out of fences) that was created for them.
### Actual behavior

They actually escape the pen every time a player disconnects and reconnects. 
### Steps to reproduce the behavior

Simple put a cow in a pen. Join, then disconnect. or maybe even restart the server

**Comment by Pokechu22** [comment]
Technically [vanilla behavior](https://bugs.mojang.com/browse/MC-2025) :laughing:

Does this only happen with fences?  Or does it happen with all blocks?  Fences only would signify that this is some kind of non-full-block specific issue, while all blocks would mean that something's generally wrong with wall collision.

**Comment by jungerm2** [comment]
I haven't tried with anything else.
However fence gates exhibit a different behavior. An attacking wolf could not pass through a open fence gate.

I'm not sure i made myself entirely clear, mobs cannot get out of the pen while you are playing. They get out when you reconnect. These aren't other mobs that spawn nearby but the same ones. I know this because i mated two cows in the pen and they had a baby. Upon reconnecting all 3 cows were out of the pen including the baby cow.

And no, in vanilla minecraft, if you put a cow in a pen and come back later (even after disconnecting) you expect the cow to still be there. No?

Hope this helps!

**Comment by onlybuilt4cubanlinx** [comment]
> And no, in vanilla minecraft, if you put a cow in a pen and come back later (even after disconnecting) you expect the cow to still be there. No?

I've never had a issue with animals leaving the pen with any kind of MC program unless it was over crowded. At that point they might get pushed/ glitch through the fence. So yes I'd expect them to be there.

**Comment by Pokechu22** [comment]
The comment about them escaping the pen being vanilla behavior was a joke :P

It's something that happens in vanilla _rarely_ which is still a bug and still annoying, but that does happen there.

**Comment by jungerm2** [comment]
So is there a work around this problem? Can i do anything to temporarily fix it or is this some sort of bug?

**Comment by Pokechu22** [comment]
It's a bug, for sure.  You _might_ be able to work around it by putting your cows in a pit rather than using a fenced pen (that's what I've done in vanilla in the past, when escaping animals was a bigger issue).

Also, I'll note that I'm unable to reproduce it on my computer using the 32-bit build on windows.  I tried with 80+ cows in a single 3x3 pen and it wasn't happening, even after restarting the server.  Can you give a bit more info about your setup?

**Comment by jungerm2** [comment]
Here's a picture of the empty pen, and then one with one of the cows that Disappeared.

I am running this on a Raspbery Pi B+ through tmux.
So it runs on Raspbian.

![2016-05-28_15 17 01](https://cloud.githubusercontent.com/assets/17375657/15629851/df48e890-24f2-11e6-9d6f-87a7710c6402.png)
![2016-05-28_15 39 51](https://cloud.githubusercontent.com/assets/17375657/15629852/df51f6d8-24f2-11e6-8b63-15f3d860e0d8.png)

**Comment by jungerm2** [comment]
For some odd reason they are all attracted to the SW corner and it seems as though hey only escape through there. This happens even with only one cow. I just saw it happening. i was flying above, the cow was in that same corner pushing against it and all of a sudden passed on the other side. I didn't have to disconnect. it might be a problem with non-full-blocks.... this didn't happen when they were in a dirt pit.

**Comment by Pokechu22** [comment]
There's now a comment on [the vanilla issue](https://bugs.mojang.com/browse/MC-2025?focusedCommentId=317274&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-317274) that suggests that this is caused by float/double rounding errors when loading the entity NBT.  I'm not completely sure if cuberite would be effected the same way, but it's possible that that's the cause here too.

**Comment by LogicParrot** [comment]
You said the cows go south-west. Are the coordinates of the location a positive X and Z? If so, the cows are potentially heading towards `0,0,0` for some reason.

**Comment by Schwertspize** [comment]
That would make sense, unless specified otherwise, rounding defaults to cutting off the edge of the number (math floor in Javascript. Just 0.7 gets 0).

**Comment by bearbin** [comment]
@jungerm2 are you confirming that the issue is fixed by closing it? If not then I will reopen - if you don't want to get notifications you can unsubscribe instead.

**Comment by jungerm2** [comment]
Yeah my bad, I closed it because it's been inactive for nearly 5 years and because it seems that this is indeed vanilla behavior. I don't know if this bug still persists.

**Comment by bearbin** [comment]
Reopening, if anybody can confirm fixed or not that would be good :)
